File: src/matlab/DynamicClient.m
classdef DynamicClient

    properties (Constant, Access = 'private')
        LIBRARY = 'grpc_dynamic_client';
        HEADER = 'exports.h'
    end

    properties (Access = 'private')
        handle
    end

    methods (Access = 'public')
        function obj = DynamicClient(target)
            if ~libisloaded(DynamicClient.LIBRARY)
                loadlibrary(DynamicClient.LIBRARY, DynamicClient.HEADER);
            end
            [~, ~, obj.handle] = calllib(DynamicClient.LIBRARY, 'Init', target, []);
        end

        function response = query(obj, service, method, request)
            [error_code, ~, ~, ~, ~, response_handle] = calllib(DynamicClient.LIBRARY, 'Query', obj.handle, service, method, request, []);
            DynamicClient.check_error(error_code);
            [~, ~, ~, size] = calllib(DynamicClient.LIBRARY, 'ReadResponse', response_handle, [], 0);
            buffer = blanks(size);
            [~, ~, response] = calllib(DynamicClient.LIBRARY, 'ReadResponse', response_handle, buffer, size);
        end

        function response = query_struct(obj, service, method, request)
            encoded_request = jsonencode(request);
            encoded_response = obj.query(service, method, encoded_request);
            response = jsondecode(encoded_response);
        end

        function close(obj)
            calllib(DynamicClient.LIBRARY, 'Close', obj.handle);
        end
    end

    methods (Static, Access = 'private')
        function check_error(error_code)
            if error_code < 0
                error(DynamicClient.get_error_message(error_code));
            end
        end

        function error_message = get_error_message(error_code)
            [~, ~, size] = calllib(DynamicClient.LIBRARY, 'GetErrorMessage', error_code, [], 0);
            buffer = blanks(size);
            [~, error_message] = calllib(DynamicClient.LIBRARY, 'GetErrorMessage', error_code, buffer, size);
        end
    end
end

File: src/dynamic_client.cpp
#include "dynamic_client.h"

#include "error_code.h"
#include "exceptions.h"
#include "session.h"
#include "unary_unary_dynamic_client.h"

using grpc::ChannelCredentials;
using ni::dynclient::DynamicClientException;
using ni::dynclient::ErrorCode;
using ni::dynclient::Session;
using ni::dynclient::UnaryUnaryDynamicClient;
using std::lock_guard;
using std::mutex;
using std::shared_ptr;
using std::string;

int32_t Init(const char* target, void** const session_handle)
{
    shared_ptr<ChannelCredentials> credentials = grpc::InsecureChannelCredentials();
    Session* session = new Session(target, credentials);
    lock_guard<mutex> _lock(session->lock());
    session->ClearLastException();
    try
    {
        session->client().QueryReflectionService();
    }
    catch (DynamicClientException ex)
    {
        session->last_exception() = ex;
    }
    return session->last_error_code();
}

int32_t Write(void* const session_handle, const char* service, const char* method, const char* request)
{
    UnaryUnaryDynamicClient* handle = (UnaryUnaryDynamicClient*)session_handle;
    try
    {
        handle->Write(service, method, request);
    }
    catch (DynamicClientException& ex)
    {
        return ex.error_code();
    }
    catch (...)
    {
        return ErrorCode::UNKNOWN;
    }
    return 0;
}

int32_t Read(void* const session_handle, char* buffer, size_t* const size)
{
    UnaryUnaryDynamicClient* handle = (UnaryUnaryDynamicClient*)session_handle;
    const string* response = handle->Read();
    if (buffer != nullptr)
    {
        response->copy(buffer, *size, 0);
    }
    else if (size != nullptr)
    {
        *size = response->size();
    }
    return 0;
}

int32_t Close(void* const session_handle)
{
    delete session_handle;
    return 0;
}

int32_t GetErrorMessage(int32_t error_code, char* const message, size_t* const size)
{
    string error_message;
    switch (error_code)
    {
    case ErrorCode::NONE:
        break;
    case ErrorCode::SERVICE_NOT_FOUND:
        error_message = "Service not found";
        break;
    case ErrorCode::METHOD_NOT_FOUND:
        error_message = "Method not found";
        break;
    default:
        error_message = "Error code is undefined";
    }

    if (message != nullptr)
    {
        error_message.copy(message, *size, 0);
    }
    else if (size != nullptr)
    {
        *size = error_message.size();
    }
    return 0;
}
